src/nvim/os/shell.c

```diff
@@ -309,15 +318,134 @@ static void system_data_cb(Stream *stream, RBuffer *buf, size_t count,
   dbuf->len += nread;
 }
 
+/// Tracks output received for the current executing shell command. To avoid
+/// flooding the UI, output is periodically skipped and a pulsing "..." is
+/// shown instead. Tracking depends on the synchronous/blocking nature of ":!".
+//
+/// Purpose:
+///   1. CTRL-C is more responsive. #1234 #5396
+///   2. Improves performance of :! (UI, esp. TUI, is the bottleneck).
+///   3. Avoids OOM during long-running, spammy :!.
+///
+/// Vim does not need this hack because:
+///   1. :! in terminal-Vim runs in cooked mode, so CTRL-C is caught by the
+///      terminal and raises SIGINT out-of-band.
+///   2. :! in terminal-Vim uses a tty (Nvim uses pipes), so commands
+///      (e.g. `git grep`) may page themselves.
+///
+/// @param size Length of data, used with internal state to decide whether
+///             output should be skipped. size=0 resets the internal state and
+///             returns the previous decision.
+///
+/// @returns true if output was skipped and pulse was displayed.
+///          Returns the previous decision if size=0.
+static bool out_data_throttle(size_t size)
+{
+  static uint64_t   started     = 0;  // Start time of the current throttle.
+  static size_t     received    = 0;  // Bytes observed since last throttle.
+  static size_t     visit       = 0;  // "Pulse" count of the current throttle.
+  static size_t     max_visits  = 0;
+  static char       pulse_msg[] = { ' ', ' ', ' ', '\0' };
+
+  if (!size) {
+    // bool previous_decision = (visit > 0);
+    started = received = visit = 0;
+    max_visits = 20;
+    return true;
+  }
+
+  received += size;
+  if (received < OUT_DATA_THRESHOLD
+      // Display at least the first chunk of output even if it is big.
+      || (!started && received < size + 1000)) {
+    return false;
+  } else if (!visit) {
+    started = os_hrtime();
+  } else if (visit >= max_visits) {
+    uint64_t since = os_hrtime() - started;
+    if (since < NS_1_SECOND) {
+      // Adjust max_visits based on the current relative performance.
+      // Each "pulse" period should last >=1 second so that it is perceptible.
+      max_visits = (2 * max_visits);
+    } else {
+      received = visit = 0;
```

Comment:
I don't reset it until the next shell command. This is to avoid calls to hrtime() which is quite slow.